src/collapse/dtrace.rs

```diff
@@ -0,0 +1,219 @@
+            count: 0,
+            occurrences: HashMap::default(),
+            pname: None,
+            opt,
+        }
+    }
+}
+
+impl Folder {
+    fn on_line(&mut self, line: &str) {
+        self.in_event = true;
+        self.on_stack_line(line)
+    }
+
+    fn stack_line_parts(line: &str) -> Option<(&str, &str)> {
+        let mut line = line.splitn(2, '+');
+        let path = line.next()?;
+
+        let addr = if let Some(addr) = line.next() {
+            addr
+        } else {
+            ""
+        };
+        Some((path, addr))
+    }
+
+    // we have a stack line that shows one stack entry from the preceeding event, like:
+    //
+    //     unix`tsc_gethrtimeunscaled+0x21
+    //     genunix`gethrtime_unscaled+0xa
+    //     genunix`syscall_mstate+0x5d
+    //     unix`sys_syscall+0x10e
+    //       1
+    fn on_stack_line(&mut self, line: &str) {
+        let line = line.trim_start();
+        if let Ok(count) = line.parse::<usize>() {
+            self.count = count;
+        } else {
+            if let Some((func, addr)) = Folder::stack_line_parts(line) {
+                self.stack.push_front(func.to_string());
+            } else {
+                warn!("weird stack line: {}", line);
+            }
+        }
+    }
+
+    fn after_event(&mut self) {
+        if !self.skip_stack && self.in_event {
+            // allocate a string that is long enough to hold the entire stack string
+            let mut len = self.stack.iter().fold(0, |a, s| a + s.len() + 1);
+            if let Some(pname) = &self.pname {
+                len += pname.len();
+            }
+            let mut stack_str = String::with_capacity(len);
+
+            // add the comm name
+            let mut first = true;
+            if let Some(pname) = &self.pname {
+                stack_str.push_str(pname);
+                first = false;
+            }
+
+            // add the other stack entries (if any)
+            for e in self.stack.drain(..) {
+                if first {
+                    first = false
+                } else {
+                    stack_str.push_str(";");
+                }
+                stack_str.push_str(&e);
+            }
+
+            // count it!
+            *self.occurrences.entry(stack_str).or_insert(0) += self.count;
+        }
+        // reset for the next event
+        self.count = 0;
+        self.in_event = false;
+        self.skip_stack = false;
+        self.stack.clear();
+    }
+
+    fn finish<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        let mut keys: Vec<_> = self.occurrences.keys().collect();
+        keys.sort();
+        for key in keys {
+            writeln!(writer, "{} {}", key, self.occurrences[key])?;
+        }
+        Ok(())
+    }
+}
+
+// massage function name to be nicer
+// NOTE: ignoring https://github.com/jvm-profiling-tools/perf-map-agent/pull/35
+fn with_module_fallback(module: &str, func: &str, pc: &str, include_addrs: bool) -> String {
```

Comment:
I don't think this function is currently used?

Reply:
👍